src/index.ts

```diff
@@ -1,3 +1,182 @@
 // Copyright (c) Jupyter Development Team.
 // Distributed under the terms of the Modified BSD License.
 'use-strict';
+
+import {
+  IInputAreaViewModel
+} from 'jupyter-js-input-area';
+
+import {
+  IOutputAreaViewModel
+} from 'jupyter-js-output-area';
+
+import {
+  IObservableList
+} from 'phosphor-observablelist';
+
+import {
+  Widget
+} from 'phosphor-widget';
+
+import './index.css';
+
+
+/**
+ * An enum which describes the type of cell.
+ */
+enum CellType {
+  /**
+   * The cell contains code input.
+   */
+  Code,
+
+  /**
+   * The cell contains markdown.
+   */
+  Markdown,
+
+  /**
+   * The cell contains raw text.
+   */
+  Raw
+}
+
+
+/**
+ * An enum which describes which interface property has changed.
+ */
+export
+enum CellProperty {
+  CellType,
+  Tags,
+  Deletable,
+  Mergable,
+  Splittable,
+  Rendered,
+  Marked,
+  Format
+}
+
+
+/**
+ * The arguments object emitted with the `stateChanged` signal.
+ */
+export
+interface ICellChangedArgs<T> {
+  which: CellProperty,
+  oldValue: T;
+  newValue: T;
+}
+
+
+/**
+ * The definition of a model object for a base cell.
+ */
+interface IBaseCellViewModel {
+
+  /**
+   * The type of cell.
+   */
+  cellType: CellType;
+
+  /**
+   * Tags applied to the cell.
+   */
+  tags?: IObservableList<string>;
+
+  /**
+   * Get namespaced metadata about the cell.
+   */
+  getMetadata(namespace: string) : IObservableMap<string, JSONData>;
```

Comment:
It's the type safety i'm worried about here. It's whether the cell/notebook **view** model is the right place for it. Are we expecting **all** implementations of `ICell` to provide this generic (meta)data store? If so, are they free to implement it in-memory only?

Otherwise, the metadata could be accessible at a lower layer of the model hierarchy, and we can expose a way to get to access those models which will be available to the cell toolbar.